cdap-security/src/test/java/co/cask/cdap/security/authorization/InMemoryAuthorizer.java

```diff
@@ -168,24 +170,32 @@ public void removeRoleFromPrincipal(Role role, Principal principal) throws RoleN
         privileges.addAll(getPrivileges(role));
       }
     }
-    return privileges;
+    return Collections.unmodifiableSet(privileges);
   }
 
   private Set<Privilege> getPrivileges(Principal principal) {
-    Set<Privilege> privileges = new HashSet<>();
-    Set<Map.Entry<EntityId, Set<Action>>> entries = table.column(principal).entrySet();
-    for (Map.Entry<EntityId, Set<Action>> entry : entries) {
-      for (Action action : entry.getValue()) {
-        privileges.add(new Privilege(entry.getKey(), action));
+    Set<Privilege> result = new HashSet<>();
+    for (Map.Entry<EntityId, ConcurrentMap<Principal, Set<Action>>> entry : privileges.entrySet()) {
+      EntityId entityId = entry.getKey();
+      ConcurrentMap<Principal, Set<Action>> value = entry.getValue();
+      Set<Action> actions = value.get(principal);
+      if (actions == null) {
+        continue;
+      }
+      for (Action action : actions) {
+        result.add(new Privilege(entityId, action));
       }
     }
-    return privileges;
+    return Collections.unmodifiableSet(result);
   }
 
   private Set<Action> getActions(EntityId entity, Principal principal) {
-    if (!table.contains(entity, principal)) {
-      table.put(entity, principal, new HashSet<Action>());
-    }
-    return table.get(entity, principal);
+    // add an entry for the specified entity, if it does not already exist
+    privileges.putIfAbsent(entity, new ConcurrentHashMap<Principal, Set<Action>>());
+    // a valid entry is now guaranteed to exist for the specified entity.
+    // now add an entry for the specified principal on the specified entity, if it does not already exist
+    privileges.get(entity).putIfAbsent(principal, Collections.newSetFromMap(new ConcurrentHashMap<Action, Boolean>()));
```

Comment:
The `privileges.get(entity)` may return `null` if the `revoke` was triggered between the `putIfAbsent` and this `get` call.